### PR TITLE
Added Typescript Definitions & Annotations

### PR DESCRIPTION
**Pull request by jacob-shuman**
This is to resolve issue #15

All files are now in .ts format and export type definition files inside of build.

Type annotations were added to existing functions and data members to enhance clarity.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -3,12 +3,12 @@
   "version": "1.6.0",
   "description": "Run IPC calls with a promise API.",
   "scripts": {
-    "pretest": "npm run --silent lint && npm run build",
+    "pretest": "npm run build && npm run --silent lint ",
```
Comment (sibnerian) [inline package.json]: Why the change here?

File: src/base.ts [hunk 2]
```diff
@@ -0,0 +1,108 @@
+ import uuid from 'uuid/v4';
+ import serializeError from 'serialize-error';
+ import Map from 'es6-map';
+ import { IpcMain, IpcRenderer, WebContents, IpcMessageEvent } from 'electron';
+//  import { ipcMain } from 'electron'; // eslint-disable-line
```
Comment (sibnerian) [inline src/base.ts]: I assume this change is because you need types for IpcRenderer, WebContents, IpcMessageEvent -- is that correct?

If so :+1:, but please delete commented line.

File: src/base.ts [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+export type Listener = (event?: IpcMessageEvent, ...dataArgs: any) => void
+export type Options = { maxTimeoutMs?: number }
+
+ export default class PromiseIpcBase {
+   public eventEmitter: IpcMain | IpcRenderer;
+   public maxTimeoutMs: number;
+   public routeListenerMap: Map;
+   public listenerMap: Map;
```
Comment (sibnerian) [inline src/base.ts]: These can be private I think. They're not part of the public API of the class.

File: src/base.ts [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+      this.maxTimeoutMs = opts.maxTimeoutMs;
+    }
+
+     // either ipcRenderer or ipcMain
+     this.eventEmitter = eventEmitter;
+     this.routeListenerMap = new Map();
+     this.listenerMap = new Map();
+   }
+
+   public send(route: string, sender: WebContents | IpcRenderer, ...dataArgs: any): Promise<void> {
```
Comment (sibnerian) [inline src/base.ts]: ❌ This is not a Promise<void>, it's a Promise<typeof returnData>. Please fix.

File: src/mainProcess.ts [hunk 5]
```diff
@@ -1,20 +1,23 @@
...
-  send(route, webContents, ...dataArgs) {
+  public send(route: string, webContents: WebContents, ...dataArgs: any): Promise<void> {
     return super.send(route, webContents, ...dataArgs);
   }
 }
 
 export const PromiseIpc = PromiseIpcMain;
 
-const mainExport = new PromiseIpcMain();
+const mainExport = new PromiseIpcMain({});
```
Comment (sibnerian) [inline src/mainProcess.ts]: I don't see a need for this change.

File: src/mainProcess.ts [hunk 6]
```diff
@@ -1,20 +1,23 @@
...
+  public PromiseIpc?: any;
+  public PromiseIpcMain?: any;
+
+  constructor(opts?: Options) {
     super(opts, ipcMain);
   }
 
   // Send requires webContents -- see http://electron.atom.io/docs/api/ipc-main/
-  send(route, webContents, ...dataArgs) {
+  public send(route: string, webContents: WebContents, ...dataArgs: any): Promise<void> {
```
Comment (sibnerian) [inline src/mainProcess.ts]: Not a `Promise<void>`

File: src/renderer.ts [hunk 7]
```diff
@@ -1,12 +1,15 @@
...
-  constructor(opts) {
+  public PromiseIpc?: any;
+  public PromiseIpcRenderer?: any;
+
+  constructor(opts?: Options) {
     super(opts, ipcRenderer);
   }
-
-  send(route, ...dataArgs) {
+
```
Comment (sibnerian) [inline src/renderer.ts]: Remove whitespace

File: src/renderer.ts [hunk 8]
```diff
@@ -1,12 +1,15 @@
 import { ipcRenderer } from 'electron'; // eslint-disable-line
-import PromiseIpcBase from './base';
+import PromiseIpcBase, { Options } from './base';
 
 export class PromiseIpcRenderer extends PromiseIpcBase {
-  constructor(opts) {
+  public PromiseIpc?: any;
```
Comment (sibnerian) [inline src/renderer.ts]: Why are these here? They don't make much sense to me. These fields didn't exist before.

File: src/mainProcess.ts [hunk 9]
```diff
@@ -1,20 +1,23 @@
-import { ipcMain } from 'electron';
-import PromiseIpcBase from './base';
+import { ipcMain, WebContents } from 'electron';
+import PromiseIpcBase, { Options } from './base';
 
 export class PromiseIpcMain extends PromiseIpcBase {
-  constructor(opts) {
+  public PromiseIpc?: any;
+  public PromiseIpcMain?: any;
```
Comment (sibnerian) [inline src/mainProcess.ts]: I don't understand why these fields are now necessary.

Comment (sibnerian) [reply src/mainProcess.ts]: Hold the phone, I just realized why you've got them. It's for this assignment to type-check right?

```ts
mainExport.PromiseIpc = PromiseIpcMain;
mainExport.PromiseIpcMain = PromiseIpcMain;
```

Comment (sibnerian) [reply src/mainProcess.ts]: I feel like this is kludgy because:

1. Not every `PromiseIpc` will have these defined
2. They're not typed; both of these should be type ?PromiseIpcMain.

So maybe the better approach here, I think, is to define an [intersection type](https://www.typescriptlang.org/docs/handbook/advanced-types.html#intersection-types). Seems like

```ts
type MainProcessExportType = PromiseIpcMain & {PromiseIpc: ?PromiseIpcMain, PromiseIpcMain: ?PromiseIpcMain};
```

And then just cast the mainExport object to that type.

File: package.json (hunk 1 above)
Comment (jacob-shuman) [reply package.json]: That was my bad sorry, for some reason I thought that the listing would be necessary prior to building. You're right this isn't very necessary.

File: src/base.ts (hunk 2 above)
Comment (jacob-shuman) [reply src/base.ts]: That was the purpose for the new imports.

Sorry about that, the commented line will be removed.

File: src/base.ts (hunk 3 above)
Comment (jacob-shuman) [reply src/base.ts]: Sounds Good!

I didn't know if any of them were intended to be public. I have privatized them in the latest commit

File: src/base.ts (hunk 4 above)
Comment (jacob-shuman) [reply src/base.ts]: Just updated it to `Promise<unknown>`

File: src/renderer.ts (hunk 7 above)
Comment (jacob-shuman) [reply src/renderer.ts]: The extra line has been removed

File: src/mainProcess.ts (hunk 6 above)
Comment (jacob-shuman) [reply src/mainProcess.ts]: Just updated to `Promise<unknown>`

File: src/mainProcess.ts (hunk 9 above)
Comment (jacob-shuman) [reply src/mainProcess.ts]: Sounds good, just updated the mainExport type!
